datashader/geo.py

```diff
@@ -0,0 +1,320 @@
+from __future__ import division
+
+import numpy as np
+import datashader.transfer_functions as tf
+
+from datashader.colors import rgb
+from datashader.utils import ngjit
+from xarray import DataArray
+
+__all__ = ['mean', 'binary', 'slope', 'aspect', 'ndvi', 'hillshade']
+
+
+def _shade(altituderad, aspect, azimuthrad, slope):
+    shade = np.sin(altituderad) * np.sin(slope) + np.cos(altituderad) * np.cos(slope) * np.cos(azimuthrad - aspect)
+    return shade
+
+def _threshold_hs(img):
+    dt = np.dtype((np.int32, {'r': (np.uint8, 0),
+                              'g': (np.uint8, 1),
+                              'b': (np.uint8, 2),
+                              'a': (np.uint8, 3)}))
+    img.data = np.where(img.data.view(dtype=dt)['r'] > 105, 0, img) # awk.
+    return img
+
+def _simple_hs(altitude, aspect, azimuth, slope, cmap, alpha, out_type='image'):
+    _shaded = _shade(altitude, aspect, azimuth, slope)
+    agg = DataArray(_shaded, dims=['x', 'y'])
+    agg.data = np.where(agg > agg.mean(), 0, agg)
+    if out_type == 'image':
+        img = tf.shade(agg, cmap=cmap, how='linear', alpha=alpha)
+        return _threshold_hs(img)
+    elif out_type == 'data':
+        return agg
+    else:
+        raise ValueError("Unknown  out_type: {0}".format(out_type))
+
+
+def _mdow_hs(altitude, aspect, azimuth, slope, cmap, alpha, out_type='image'):
+    alt = np.deg2rad(30)
+    shade = np.sum([_shade(alt, aspect, np.deg2rad(225), slope),
+                    _shade(alt, aspect, np.deg2rad(270), slope),
+                    _shade(alt, aspect, np.deg2rad(315), slope),
+                    _shade(alt, aspect, np.deg2rad(360), slope)], axis=0)
+
+#     shade = Normalize()(shade)
+    agg = DataArray(shade, dims=['y', 'x'])
+    if out_type == 'image':
+        img = tf.shade(agg, cmap=cmap, how='linear', alpha=alpha)
+        return _threshold_hs(img)
+    elif out_type == 'data':
+        return agg
+    else:
+        raise ValueError("Unknown  out_type: {0}".format(out_type))
+
+
+def _swiss_hs(altitude, aspect, azimuth, slope, cmap, alpha):
+    pass
+
+_hillshade_lookup = {'simple': _simple_hs,
+                     'mdow': _mdow_hs,
+                     'swiss': _swiss_hs}
+
+def _normalize_hillshade_how(how):
+    if how in _hillshade_lookup:
+        return _hillshade_lookup[how]
+    raise ValueError("Unknown hillshade method: {0}".format(how))
+
+def hillshade(agg,
+              altitude=30,
+              azimuth=315,
+              alpha=70,
+              how='simple',
+              out_type='image',
+              cmap=['#C7C7C7', '#000000']):
+    """Convert a 2D DataArray to an hillshaded image with specified colormap.
+
```

Comment:
Can you include a citation or link to provide a canonical reference or specification for this algorithm?